pygsti/circuits/circuit.py

```diff
@@ -3744,22 +3749,200 @@ def convert_to_cirq(self,
             gatename_conversion = _itgs.standard_gatenames_cirq_conversions()
             if wait_duration is not None:
                 gatename_conversion[idle_gate_name] = cirq.WaitGate(wait_duration)
+        #conversion does not work is the line labels are none, or the line labels are not a subset
+        #of the keys for qubit_conversion (indicating there isn't a corresponding mapping into cirq objects).
+        msg1 = 'Conversion to cirq does not work with circuits w/placeholder * line label.'
+        msg2 = 'Missing qubit conversions, some line labels have no corresponding cirq conversion in qubit_conversions.'
+        assert self.line_labels != ('*',), msg1
+        assert set(self.line_labels).issubset(set(qubit_conversion.keys())), msg2
 
         moments = []
         for i in range(self.num_layers):
             layer = self.layer_with_idles(i, idle_gate_name)
             operations = []
             for gate in layer:
                 operation = gatename_conversion[gate.name]
-                if operation is None:
-                    # This happens if no idle gate it specified because
-                    # standard_gatenames_cirq_conversions maps 'Gi' to `None`
-                    continue
                 qubits = map(qubit_conversion.get, gate.qubits)
                 operations.append(operation.on(*qubits))
             moments.append(cirq.Moment(operations))
 
         return cirq.Circuit(moments)
+    
+    @classmethod
+    def from_cirq(cls, circuit, qubit_conversion=None, cirq_gate_conversion= None,implied_idles = False, global_idle = False):
+        """
+        Converts and instantiates a pyGSTi Circuit object from a Cirq Circuit object.
+
+        Parameters
+        ----------
+        circuit : cirq Circuit
+            The cirq Circuit object to parse into a pyGSTi circuit.
+
+        qubit_conversion : dict, optional (default None)
+            A dictionary specifying a mapping between cirq qubit objects and 
+            pyGSTi qubit labels (either integers or strings).
+            If None, then a default mapping is created.
+
+        cirq_gate_conversion : dict, optional (default None)
+            If specified a dictionary with keys given by cirq gate objects,
+            and values given by pygsti gate names which overrides the built-in
+            conversion dictionary used by default.
+
+        implied_idles : bool, optional (default False)
+            A flag indicating whether to explicitly include
+            implied idles as part of a circuit layer containing
+            other explicitly specified gates.
+
+        global_idle : bool or string or Label, optional (default False)
```

Comment:
A question/clarification on implicit vs global idle behavior. The default is False for both, but this results in Gi's being dropped for implicit idles everywhere EXCEPT for global idles - that is, in cell 17 of the example, we have Gi labels in partial idle layers, but the global idle layer does have Gis. I kinda expected no Gis anywhere in this case.

Maybe more succintly put, I expected the "True" behavior of global idle to be the default (i.e. cell 20), and to recover the current (i.e. cell 17) behavior, a user would have to explicitly put "Gi" for the global_idle string here. Another advantage of this approach is then the global_idle argument only takes one type of argument (a label-like one) rather than switching behavior on boolean/label-like arguments.

I am open to being convinced that current behavior is most desired though!

P.S. Another place where Jupytext examples/tutorials would probably make this discussion easier...

Reply:
Yeah, I was conflicted on the default behavior here. The original decision to make replacement of the global idle layers non-default is that unlike with implicit idles where the convention is pretty universally accepted (i.e. not including them), the convention for global idles is less so. I know we have Label(()), but I've always kind of disliked this convention (from a dev perspective it is kind of annoying and creates lots of edge cases to check for), so I initially wanted to force the choice of substitution to be an affirmative one. As you can see from the final implementation I eventually flip-flopped on this and did add a default label replacement choice of Label(()), so I think making this default would be alright. I'll probably change the names of these kwargs to be more descriptive while I'm at it.